### PR TITLE
Do not use __managed__ in GPU builds. Add utils.

### PR DESCRIPTION
**Pull request by olupton**
**Description**
This fixes CPU execution of GPU builds on machines that do not have GPUs, which were previously segfaulting due to the use of the `__managed__` keyword. Now the handling of Random123 state is more explicit for host/device. This was only added in https://github.com/BlueBrain/CoreNeuron/pull/595, but the issue wasn't noticed at the time because the GPU-enabled CI only executes tests on machines with GPUs.

Also add a pair of helper functions `coreneuron::[de]allocate_unified()` that wrap `cudaMallocManaged()` in GPU builds if `--gpu` was passed at runtime and fall back to `new`/`delete` otherwise, and a method `coreneuron::unified_memory_enabled()` that queries whether this condition is met.

Additionally add a C++ allocator template `coreneuron::unified_allocator<T>` that wraps these functions, and a templated `coreneuron::alloc_deleter<T>` for use with `std::unique_ptr<T, D>`. Also add `coreneuron::allocate_unique` helper from SO.

Cleanup Random123 code by dropping an unused `nrnran123_mutconstruct()` method.

Tweak compilation scripts to allow for circular dependencies between `libcoreneuron.a` and `libcudacoreneuron.a`.
In future these should just be merged into a single library.

This addresses https://github.com/BlueBrain/CoreNeuron/issues/599#issuecomment-888416482. #599 should stay open because various OpenACC calls are still not conditional on `--gpu`.

**How to test this?**
Try running a GPU-built `special-core` without `--gpu` on a machine that does not have an NVIDIA GPU.

**Test System**
 - OS: BB5
 - Compiler: NVHPC 21.7 / CUDA 11.0 / GCC 9.3
 - Version: master
 - Backend: GPU/CPU

**Use certain branches for the SimulationStack CI**

CI_BRANCHES:NEURON_BRANCH=master,

### CODE REVIEW

File: CMake/OpenAccHelper.cmake
Comment (ferdonline) [inline CMake/OpenAccHelper.cmake:72]: I see `-lcoreneuron` twice. Expected?

Comment (olupton) [reply CMake/OpenAccHelper.cmake:72]: Yes, although I haven't actually tested the `nrnivmodl` workflow (only `nrnivmodl-core`) where IIUC this is used -- the CI will do that for us...

It's a crude workaround (commented elsewhere) for circular dependencies between `libcoreneuron.a` and `libcudacoreneuron.a` introduced by this PR. I didn't immediately find a better **simple** way.

IMO the better solution would be for us to produce a single library instead of these two separate ones. I don't think they are really logically separate libraries, the separation is just a consequence of how things are handled in CMake.

Alternatively we could add a third library...
```
libcoreneuron.a (C++) <depends on> libcudacoreneuron.a (CUDA) <depends on> libcorecoreneuron.a (C++)
```
but I fear this will just get messier without much real benefit.

**Comment by olupton** [comment]
This is blocked by #607.

File: CMake/OpenAccHelper.cmake
Comment (olupton) [reply CMake/OpenAccHelper.cmake]: #609 made it easy to merge `libcoreneuron` and `libcudacoreneron`, so I did that instead. Much neater.

**Review by ferdonline — COMMENTED** [review]
LGTM but I'm not the best person to review this code. Maybe let's wait for a review from @iomaganaris

**Review by kotsaloscv — APPROVED** [review]
LGTM

File: coreneuron/utils/randoms/nrnran123.cu
Comment (pramodk) [inline coreneuron/utils/randoms/nrnran123.cu:224]: Answered in previous comment.

@nrnhines : How the ran123 streams allocated with `nrnran123_newstream3` inside `bbcore_read()` should be allocated? any thoughts?

File: coreneuron/utils/randoms/nrnran123.cu
Comment (olupton) [reply coreneuron/utils/randoms/nrnran123.cu:185]: There wasn't any discussion, but there was a pre-existing comment:
https://github.com/BlueBrain/CoreNeuron/blob/170a0bb74b033de020946c96f67f8699e67e10d4/coreneuron/utils/randoms/nrnran123.cu#L48
which made me think that  we could easily add a diagnostic for this. I haven't tried to run the `psolve-direct` tests with this branch. You're probably right that this should be introduced as a warning instead of a hard error.

File: coreneuron/CMakeLists.txt
Comment (olupton) [reply coreneuron/CMakeLists.txt]: Removed.

File: coreneuron/utils/randoms/nrnran123.cu
Comment (olupton) [reply coreneuron/utils/randoms/nrnran123.cu]: I demoted this to a warning.

File: coreneuron/utils/randoms/nrnran123.cu
Comment (pramodk) [reply coreneuron/utils/randoms/nrnran123.cu:223]: lets keep this as a separate issue. https://github.com/BlueBrain/nmodl/issues/383 would help to implement this easily.

**Review by pramodk — APPROVED** [review]
LGTM

**Pull request merged**
